keps/sig-architecture/1659-standard-topology-labels/README.md

```diff
@@ -0,0 +1,294 @@
+hard-coded and not extensible.  Those labels have relatively well understood
+meanings, and (so far) have been sufficient to represent what most people need.
+
+This KEP proposes to document those labels as "standard" and give them more
+rigorous definitions.  This also proposes that we discuss and decide whether a
+third level of topology is needed and if so, define it in the same manner as
+the existing labels.
+
+The resulting definitions should be specific enough that users and implementors
+understand what they mean, but not so rigid that they can not map them to the
+nearest constructs available in most environments.
+
+### Risks and Mitigations
+
+The primary risks here are:
+
+1) That we define these too loosely, such that users can not derive sufficient
+value from their use.
+
+2) That we define these too specifically, such that implementors can not use
+them to represent natural concepts in their environents.
+
+3) That we define these in a way that is incompatible with the ways they are
+alredy being used.
+
+4) That we preclude or design-out other uses of topology that users are using
+today.
+
+## Design Details
+
+### Reserve a label prefix
+
+Label prefixes allow us to group labels on common origin and meaning.  We
+propose to document somewhere (TBD) that the prefix "topology.kuberntes.io" is
+explicitly reserved for use in defining metadata about the physical or logical
+connectivity and grouping of Kubernetes nodes, and the associated behavioral
+and failure properties of those groups.
+
+This prefix is already in use.  This KEP just aims to formalize it.
+
+### Defining the meaning of existing labels
+
+This KEP proposes to define the meaning and semantics of the following labels:
+
+    * topology.kubernetes.io/region
+    * topology.kubernetes.io/zone
+
+The exact wording is TBD, but it must be specific enough to be useful to users
+and loose enough to allow implementors sufficient freedom.
+
+This will also include defining that "region" and "zone" are strictly
+hierarchical ("zones" are subsets of "regions") and that zone names are unique
+across regions.  For example AWS documents "us-east-1a" as a zone under region
+"us-east-1".
+
+<<[UNRESOLVED]>>
+Should we also try to standardize "kubernetes.io/hostname" as "topology.kubernetes.io/node" ?
+<<[/UNRESOLVED]>>
+
+### Defining a third key (or not)
+
+Some systems define topology in two levels (e.g. public clouds) and others use three
+levels (e.g. Envoy adds "sub-zone").  This KEP proposes that we standardize on
+two levels for now, while reserving the right to expand that to three (or more)
+if and when we have strong demand.
+
+### Followup work (or optionally part of this)
+
+For a Pod to know its own topology today, it must be authorized to look at
+Nodes.  This is somewhat tedious, when we have downward-API support for labels
+already, and we know that these topology labels are not likely to change at
```

Comment:
This was actually the biggest reason for closing https://github.com/kubernetes/enhancements/pull/1127, some topologies can change at runtime (e.g. physical host topology with hypervisor doing a live migration of VMs). Should we also document that topologies should be immutable? I think this has only been true today because of what's implied with zone/region in public clouds but it's not documented AFAICT.

Reply:
I think that would be a great addition, actually.  I'll add words

Reply:
Added words